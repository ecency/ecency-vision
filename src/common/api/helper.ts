import defaults from "../constants/defaults.json";

export const apiBase = (endpoint: string): string => `${defaults.base}${endpoint}`;

export const apiBaseImage = (endpoint: string): string => `${defaults.imageServer}${endpoint}`;
