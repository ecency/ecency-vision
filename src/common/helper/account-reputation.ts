const isHumanReadable = (input: number): boolean => {
  return Math.abs(input) > 0 && Math.abs(input) <= 100;
};

export default (input: string | number): number => {
  if (typeof input === 'number' && isHumanReadable(input)) {
    return Math.floor(input);
  }

  if (typeof input === 'string') {
    input = Number(input);

    if (isHumanReadable(input)) {
      return Math.floor(input);
    }
  }

  if (input === 0) {
    return 25;
  }

  let neg = false;

  if (input < 0) neg = true;

  let reputationLevel = Math.log10(Math.abs(input));
  reputationLevel = Math.max(reputationLevel - 9, 0);

  if (reputationLevel < 0) reputationLevel = 0;

  if (neg) reputationLevel *= -1;

  reputationLevel = reputationLevel * 9 + 25;

  return Math.floor(reputationLevel);
};
