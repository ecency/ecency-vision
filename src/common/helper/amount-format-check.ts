export default (v: string) => /^\d{1,3}(,?\d{3})*(\.(\d{3},?)*\d{1,3})?$/.test(v);
