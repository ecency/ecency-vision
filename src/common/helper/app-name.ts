import stripTags from "../util/strip-tags";

interface ObjInput {
  name?: string;
}

export default (input: string | null | undefined | ObjInput) => {
  if (!input) {
    return "";
  }

  if (typeof input === "string") {
    return stripTags(input);
  }

  if (typeof input === "object" && input.name !== undefined) {
    return stripTags(input.name);
  }

  return "";
};
