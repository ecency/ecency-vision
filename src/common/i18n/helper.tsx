import React from "react";

import xss from "xss";

import { _t } from "./index";

export const HTML_WHITELIST = {
  a: ["class", "href", "target", "rel"],
  span: ["class"],
  h3: [],
  p: [],
  strong: [],
  b: [],
  img: ["src", "alt"],
  ul: [],
  li: [],
  br: []
};

export const safeHtml = (html: string): string => {
  return xss(html, {
    whiteList: HTML_WHITELIST,
    stripIgnoreTagBody: true,
    stripIgnoreTag: true,
    css: false
  });
};

interface Props {
  k: string;
  args?: {};
  children: JSX.Element;
}

export class Tsx extends React.Component<Props> {
  render() {
    const { k, args } = this.props;

    const { children } = this.props;

    const html = safeHtml(_t(k, args));

    return React.cloneElement(children, { dangerouslySetInnerHTML: { __html: html } });
  }
}
