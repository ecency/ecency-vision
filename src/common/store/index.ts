import { combineReducers } from "redux";
import { connectRouter } from "connected-react-router";
import { createBrowserHistory, createMemoryHistory, History } from "history";

import isEqual from "react-fast-compare";

import global from "./global";
import dynamicProps from "./dynamic-props";
import trendingTags from "./trending-tags";
import entries from "./entries";
import accounts from "./accounts";
import communities from "./communities";
import transactions from "./transactions";
import users from "./users";
import activeUser from "./active-user";
import reblogs from "./reblogs";
import discussion from "./discussion";
import ui from "./ui";
import subscriptions from "./subscriptions";
import notifications from "./notifications";
import points from "./points";
import signingKey from "./signing-key";
import entryPinTracker from "./entry-pin-tracker";
import persistentPageScroll from "./persistent-page-scroll";
import deck from "./deck";

import filterTagExtract from "../helper/filter-tag-extract";

let reducers = {
  global,
  dynamicProps,
  trendingTags,
  entries,
  accounts,
  communities,
  transactions,
  users,
  activeUser,
  reblogs,
  discussion,
  ui,
  subscriptions,
  notifications,
  points,
  signingKey,
  entryPinTracker,
  deck,
  persistentPageScroll
};

export let history: History | undefined;

// create browser history on client side
if (typeof window !== "undefined") {
  if (window.location.href.startsWith("file://")) {
    // electron
    history = createMemoryHistory();
  } else {
    // web
    history = createBrowserHistory();
  }

  // We need a customised history object since history pushes new state for same path.
  // See: https://github.com/ReactTraining/history/issues/470
  // We don't want LOCATION_CHANGE triggered for same path because of structure of out "entries" reducer.

  // get ref of current push function
  const _push = history.push;

  let prevPath: string = history.location.pathname;
  // update previous path once history change
  history.listen((location) => {
    prevPath = location.pathname;
  });

  // create a new push function that compares new path and previous path.
  history.push = (pathname: History.Path, state: History.LocationState = {}) => {
    // compare filter & tag resolution of paths
    // this control required because "/" == "/hot" and "/@username" == "/@username/posts"
    const ftCur = filterTagExtract(pathname);
    const ftPrev = filterTagExtract(prevPath);

    if (ftCur && ftPrev) {
      if (isEqual(ftCur, ftPrev)) {
        return;
      }
    }

    // simple path compare
    if (pathname === prevPath) {
      return;
    }
    _push(pathname.includes("//") ? "/" : pathname, state);
  };

  // scroll to top on every push action
  history.listen((location, action) => {
    if (action === "PUSH") {
      // Don't scroll to top with anchor links
      if (history!.location.hash !== "") {
        return;
      }

      setTimeout(() => {
        window.scrollTo({
          top: 0
        });
      }, 100);
    }
  });

  // @ts-ignore
  reducers = { router: connectRouter(history), ...reducers };
}

const rootReducer = combineReducers(reducers);

export default rootReducer;

export type AppState = ReturnType<typeof rootReducer>;
