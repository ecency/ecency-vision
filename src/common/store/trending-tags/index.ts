import {Dispatch} from 'redux';

import {getTrendingTags} from '../../api/hive';

import {AppState} from '../index';

import {
  TrendingTags,
  Actions,
  ActionTypes,
  FetchAction,
  FetchedAction,
  FetchErrorAction,
} from './types';

export const initialState: TrendingTags = {
  list: [],
  loading: false,
  error: false,
};

export default (
  state: TrendingTags = initialState,
  action: Actions,
): TrendingTags => {
  switch (action.type) {
    case ActionTypes.FETCH: {
      return {
        list: [],
        loading: true,
        error: false,
      };
    }
    case ActionTypes.FETCHED: {
      return {
        list: action.tags,
        loading: false,
        error: false,
      };
    }
    case ActionTypes.FETCH_ERROR: {
      return {
        list: [],
        loading: false,
        error: true,
      };
    }
    default:
      return state;
  }
};

/* Actions */
export const fetchTrendingTags =
  () => (dispatch: Dispatch, getState: () => AppState) => {
    const {trendingTags} = getState();

    if (trendingTags.list.length >= 1 || trendingTags.loading) {
      return;
    }

    dispatch(fetchAct());

    return getTrendingTags()
      .then(tags => {
        dispatch(fetchedAct(tags));
        return tags;
      })
      .catch(() => {
        dispatch(fetchErrorAct());
      });
  };

/* Action Creators */
export const fetchAct = (): FetchAction => {
  return {
    type: ActionTypes.FETCH,
  };
};

export const fetchedAct = (tags: string[]): FetchedAction => {
  return {
    type: ActionTypes.FETCHED,
    tags,
  };
};

export const fetchErrorAct = (): FetchErrorAction => {
  return {
    type: ActionTypes.FETCH_ERROR,
  };
};
