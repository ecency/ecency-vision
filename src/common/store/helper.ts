import {Store} from 'redux';

import i18n from 'i18next';

import {AppState} from './index';
import {ActiveUser, UserPoints} from './active-user/types';
import {
  loginAct as loginActiveUser,
  logoutAct as logoutActiveUser,
  updateAct as updateActiveUserAct,
} from './active-user';

import {getAccount, getDynamicProps} from '../api/hive';
import {getPoints, usrActivity, getPromotedEntries} from '../api/private-api';
import {reloadAct as reloadUsers} from './users';
import {fetchedAct as loadDynamicProps} from './dynamic-props';
import {fetchedAct as entriesFetchedAct} from './entries';
import {
  setCurrencyAct as setCurrency,
  muteNotificationsAct as muteNotifications,
  setLangAct as setLang,
  setNsfwAct as setNsfw,
} from './global';

import {getCurrencyRate} from '../api/misc';

import currencies from '../constants/currencies.json';

import * as ls from '../../common/util/local-storage';

import currencySymbol from '../../common/helper/currency-symbol';

import {AppWindow} from '../../client/window';

declare var window: AppWindow;

export const activeUserMaker = (
  name: string,
  points: string = '0.000',
  uPoints: string = '0.000',
): ActiveUser => {
  return {
    username: name,
    data: {name},
    points: {
      points,
      uPoints,
    },
  };
};

export const activeUserUpdater = async (store: Store<AppState>) => {
  const state = store.getState();
  if (state.activeUser) {
    const {username} = state.activeUser;

    let account;
    try {
      account = await getAccount(username);
    } catch (e) {
      return;
    }

    let points: UserPoints;
    try {
      const p = await getPoints(username);
      points = {points: p.points, uPoints: p.unclaimed_points};
    } catch (e) {
      points = {
        points: '0.000',
        uPoints: '0.000',
      };
    }

    store.dispatch(updateActiveUserAct(account, points));
  }
};

export const syncActiveUser = (store: Store<AppState>) => {
  const state = store.getState();

  const activeUser = ls.get('active_user');

  // logout
  if (!activeUser && state.activeUser) {
    store.dispatch(logoutActiveUser());
    return;
  }

  // login
  if (activeUser && !state.activeUser) {
    store.dispatch(loginActiveUser());
    activeUserUpdater(store).then();
    return;
  }

  // switch
  if (
    activeUser &&
    state.activeUser &&
    activeUser !== state.activeUser.username
  ) {
    store.dispatch(loginActiveUser());
    activeUserUpdater(store).then();
    return;
  }
};

export const clientStoreTasks = (store: Store<AppState>) => {
  // To use in places where we can't access to store
  window.usePrivate = store.getState().global.usePrivate;

  // Initial state from browser's local storage
  store.dispatch(reloadUsers());
  store.dispatch(loginActiveUser());

  // Load dynamic props
  getDynamicProps().then(resp => {
    store.dispatch(loadDynamicProps(resp));
  });

  // Update active user in interval
  activeUserUpdater(store).then();
  setInterval(() => {
    activeUserUpdater(store).then();
  }, 60 * 1000);

  // Active user sync between tabs / windows
  setInterval(() => {
    syncActiveUser(store);
  }, 2000);

  // Do check-in in interval
  const checkIn = () => {
    const state = store.getState();
    if (state.activeUser) {
      usrActivity(state.activeUser?.username!, 10).then();
    }
  };

  // wait for initial user sync
  setTimeout(() => {
    checkIn();
  }, 5000);
  setInterval(checkIn, 1000 * 60 * 15 + 8);

  // Inject / update promoted entries to store
  const promotedEntries = () => {
    getPromotedEntries().then(r => {
      store.dispatch(entriesFetchedAct('__promoted__', r, false));
    });
  };
  promotedEntries();
  setInterval(promotedEntries, 1000 * 60 * 5);

  // Currency
  const currency = ls.get('currency');
  if (currency && currencies.find(x => x.id === currency)) {
    const symbol = currencySymbol(currency);
    getCurrencyRate(currency).then(rate => {
      store.dispatch(setCurrency(currency, rate, symbol));
    });
  }

  // Notifications
  if (ls.get('notifications') === false) {
    store.dispatch(muteNotifications());
  }

  // Language
  const lang = ls.get('lang');
  if (lang) {
    if (store.getState().global.lang !== lang) {
      i18n.changeLanguage(lang).then(() => {
        store.dispatch(setLang(lang));
      });
    }
  }

  // NSFW
  if (ls.get('nsfw') === true) {
    store.dispatch(setNsfw(true));
  }
};
