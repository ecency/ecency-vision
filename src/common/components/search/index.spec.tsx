import React from 'react';

import {Search} from './index';
import renderer from 'react-test-renderer';
import {createBrowserHistory, createLocation} from 'history';

import {initialState as trendingTags} from '../../store/trending-tags';
import {globalInstance} from '../../helper/test-helper';

const props = {
  history: createBrowserHistory(),
  location: createLocation({}),
  global: globalInstance,
  trendingTags,
  fetchTrendingTags: () => {},
};

const component = renderer.create(<Search {...props} />);

it('(1) Default render', () => {
  expect(component.toJSON()).toMatchSnapshot();
});

it('(2) With query', () => {
  const instance: any = component.getInstance();
  instance.setState({query: 'foo'});
  expect(component.toJSON()).toMatchSnapshot();
});
