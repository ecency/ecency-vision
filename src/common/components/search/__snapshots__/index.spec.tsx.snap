// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(1) Default render 1`] = `
<div
  className="suggestion relative"
>
  <div
    className="search-box"
  >
    <div
      className="ecency-input-group flex items-stretch w-full [&>input]:rounded-[0] [&>input:first-child]:rounded-l-full [&>input:last-child]:rounded-r-full [&>.ecency-input-group-part>button]:rounded-tl-none [&>.ecency-input-group-part>button]:rounded-bl-none [&>.ecency-input-group-part>button]:h-[2.75rem]"
    >
      <div
        className="ecency-input-group-part ecency-input-group-prepend flex items-center justify-center border-r-0 rounded-tl-full rounded-bl-full bg-gray-200 dark:bg-gray-600 dark:border-gray-600 [&>svg]:w-4 [&>svg]:h-4 px-2 border-2"
        onClick={[Function]}
      >
        <svg
          fill="none"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M14 5H20"
            opacity="0.4"
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
          />
          <path
            d="M14 8H17"
            opacity="0.4"
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
          />
          <path
            d="M21 11.5C21 16.75 16.75 21 11.5 21C6.25 21 2 16.75 2 11.5C2 6.25 6.25 2 11.5 2"
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
          />
          <path
            d="M22 22L20 20"
            opacity="0.4"
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
          />
        </svg>
      </div>
      <input
        autoComplete="off"
        className="border-2 rounded-3xl py-2 px-3 w-full outline-none shadow-0 focus:border-gray-500 hover:border-gray-300 duration-300 dark:border-gray-700 dark:hover:border-gray-600 dark:bg-gray-800 aria-invalid:border-red"
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="Search 10,000,000 posts"
        type="text"
        value=""
      />
    </div>
  </div>
  <div />
</div>
`;

exports[`(2) With query 1`] = `
<div
  className="suggestion relative"
>
  <div
    className="search-box"
  >
    <div
      className="ecency-input-group flex items-stretch w-full [&>input]:rounded-[0] [&>input:first-child]:rounded-l-full [&>input:last-child]:rounded-r-full [&>.ecency-input-group-part>button]:rounded-tl-none [&>.ecency-input-group-part>button]:rounded-bl-none [&>.ecency-input-group-part>button]:h-[2.75rem]"
    >
      <div
        className="ecency-input-group-part ecency-input-group-prepend flex items-center justify-center border-r-0 rounded-tl-full rounded-bl-full bg-gray-200 dark:bg-gray-600 dark:border-gray-600 [&>svg]:w-4 [&>svg]:h-4 px-2 border-2"
        onClick={[Function]}
      >
        <svg
          fill="none"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M14 5H20"
            opacity="0.4"
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
          />
          <path
            d="M14 8H17"
            opacity="0.4"
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
          />
          <path
            d="M21 11.5C21 16.75 16.75 21 11.5 21C6.25 21 2 16.75 2 11.5C2 6.25 6.25 2 11.5 2"
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
          />
          <path
            d="M22 22L20 20"
            opacity="0.4"
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
          />
        </svg>
      </div>
      <input
        autoComplete="off"
        className="border-2 rounded-3xl py-2 px-3 w-full outline-none shadow-0 focus:border-gray-500 hover:border-gray-300 duration-300 dark:border-gray-700 dark:hover:border-gray-600 dark:bg-gray-800 aria-invalid:border-red"
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="Search 10,000,000 posts"
        type="text"
        value="foo"
      />
    </div>
  </div>
  <div />
</div>
`;
